File: 2022/day20/clojure/solution.clj
#!/usr/bin/env bb
;; Day 20: Grove Positioning System

(ns solution
  (:require [clojure.string :as str]))

(defn parse-input [text]
  (->> (str/split-lines (str/trim text))
       (mapv parse-long)))

(defn find-position [indexed orig-idx]
  "Find the current position of element with original index orig-idx."
  (loop [pos 0]
    (if (= (first (indexed pos)) orig-idx)
      pos
      (recur (inc pos)))))

(defn mix-once [indexed n]
  "Perform one round of mixing on the indexed vector."
  (loop [indexed indexed
         orig-idx 0]
    (if (>= orig-idx n)
      indexed
      (let [curr-pos (find-position indexed orig-idx)
            [_ val] (indexed curr-pos)
            ;; Remove element from current position
            without (into (subvec indexed 0 curr-pos)
                          (subvec indexed (inc curr-pos)))
            ;; Calculate new position modulo (n-1) since element is removed
            new-pos (mod (+ curr-pos val) (dec n))
            ;; Handle negative modulo properly
            new-pos (if (neg? new-pos)
                      (+ new-pos (dec n))
                      new-pos)
            ;; Insert at new position
            with-element (into (subvec without 0 new-pos)
                               (cons [orig-idx val]
                                     (subvec without new-pos)))]
        (recur with-element (inc orig-idx))))))

(defn mix [numbers times]
  "Mix the list of numbers the specified number of times."
  (let [n (count numbers)
        ;; Create indexed pairs [original-index, value]
        indexed (vec (map-indexed vector numbers))]
    (loop [indexed indexed
           t 0]
      (if (>= t times)
        (mapv second indexed)
        (recur (mix-once indexed n) (inc t))))))

(defn grove-coordinates [mixed]
  "Find sum of 1000th, 2000th, 3000th values after 0."
  (let [n (count mixed)
        zero-idx (.indexOf mixed 0)]
    (reduce + (map #(mixed (mod (+ zero-idx %) n)) [1000 2000 3000]))))

(defn part1 [text]
  "Mix once and find grove coordinates."
  (let [numbers (parse-input text)
        mixed (mix numbers 1)]
    (grove-coordinates mixed)))

(defn part2 [text]
  "Multiply by decryption key, mix 10 times."
  (let [numbers (parse-input text)
        decryption-key 811589153
        scaled (mapv #(* % decryption-key) numbers)
        mixed (mix scaled 10)]
    (grove-coordinates mixed)))

(defn -main []
  (let [script-dir (-> *file*
                       (java.io.File.)
                       (.getParentFile)
                       (.getCanonicalPath))
        input-file (str script-dir "/../input.txt")
        text (slurp input-file)]
    (println "Part 1:" (part1 text))
    (println "Part 2:" (part2 text))))

(-main)

File: 2022/day20/php/solution.php
#!/usr/bin/env php
<?php

/**
 * Day 20: Grove Positioning System
 *
 * Circular list mixing where numbers move by their value.
 */

function parseInput(string $text): array {
    return array_map('intval', array_filter(explode("\n", trim($text)), 'strlen'));
}

function mix(array $numbers, int $times = 1): array {
    $n = count($numbers);

    // Store [original_index, value] pairs
    $indexed = [];
    foreach ($numbers as $i => $val) {
        $indexed[] = [$i, $val];
    }

    for ($round = 0; $round < $times; $round++) {
        for ($origIdx = 0; $origIdx < $n; $origIdx++) {
            // Find current position of this element
            $currPos = 0;
            foreach ($indexed as $pos => $pair) {
                if ($pair[0] === $origIdx) {
                    $currPos = $pos;
                    break;
                }
            }

            // Get the value
            $val = $indexed[$currPos][1];

            // Remove from current position
            array_splice($indexed, $currPos, 1);

            // Calculate new position (modulo n-1 because we removed the element)
            // PHP modulo can return negative, so we need to handle that
            $newPos = ($currPos + $val) % ($n - 1);
            if ($newPos < 0) {
                $newPos += ($n - 1);
            }

            // Insert at new position
            array_splice($indexed, $newPos, 0, [[[$origIdx, $val]]]);
            // Fix the nested array from array_splice
            $indexed[$newPos] = [$origIdx, $val];
        }
    }

    return array_map(function($pair) { return $pair[1]; }, $indexed);
}

function groveCoordinates(array $mixed): int {
    $n = count($mixed);
    $zeroIdx = array_search(0, $mixed, true);

    $sum = 0;
    foreach ([1000, 2000, 3000] as $offset) {
        $sum += $mixed[($zeroIdx + $offset) % $n];
    }
    return $sum;
}

function part1(string $text): int {
    $numbers = parseInput($text);
    $mixed = mix($numbers, 1);
    return groveCoordinates($mixed);
}

function part2(string $text): int {
    $numbers = parseInput($text);
    $decryptionKey = 811589153;
    $numbers = array_map(function($n) use ($decryptionKey) {
        return $n * $decryptionKey;
    }, $numbers);
    $mixed = mix($numbers, 10);
    return groveCoordinates($mixed);
}

function main(): void {
    $scriptDir = dirname(__FILE__);
    $inputFile = $scriptDir . '/../input.txt';
    $text = file_get_contents($inputFile);

    echo 'Part 1: ' . part1($text) . PHP_EOL;
    echo 'Part 2: ' . part2($text) . PHP_EOL;
}

main();
